midenc-compile/src/stages/cross_ctx/lower_imports.rs

```diff
@@ -0,0 +1,281 @@
+    // get or create the module for the interface
+    let lowering_module_id =
+        Symbol::intern(format!("lower-imports-{}", import_func_id.module.as_str()));
+    // dbg!(&lowering_module_id);
+    let mut module_builder = component_builder.module(lowering_module_id);
+
+    let import_lowered_sig = flatten_function_type(high_func_ty, FlatteningDirection::Lower)
+        .map_err(|e| {
+            let message = format!(
+                "Miden CCABI import lowering generation. Signature for imported function {} \
+                 requires flattening. Error: {}",
+                import_func_id.function, e
+            );
+            diagnostics.diagnostic(Severity::Error).with_message(message).into_report()
+        })?;
+
+    if needs_transformation(&import_lowered_sig) {
+        let message = format!(
+            "Miden CCABI export lowering generation. Signature for exported function {} requires \
+             lowering. This is not supported yet.",
+            import_func_id
+        );
+        return Err(diagnostics.diagnostic(Severity::Error).with_message(message).into_report());
+    }
+    assert_core_wasm_signature_equivalence(&import_func_sig, &import_lowered_sig);
+
+    let mut builder = module_builder.function(import_func_id.function, import_func_sig)?;
+    let entry = builder.current_block();
+    let params = builder.block_params(entry).to_vec();
+
+    let dfg = builder.data_flow_graph_mut();
+    if dfg.get_import(&import_func_id).is_none() {
+        dfg.import_function(
+            import_func_id.module,
+            import_func_id.function,
+            import_lowered_sig.clone(),
+        )
+        .map_err(|_e| {
+            let message = format!(
+                "lowering function with name {} in module {} with signature \
+                 {import_lowered_sig:?} is already imported (function call) with a different \
+                 signature",
+                import_func_id.function, import_func_id.module
+            );
+            diagnostics.diagnostic(Severity::Error).with_message(message).into_report()
+        })?;
+    }
+    let call = builder.ins().call(import_func_id, &params, span);
+    // dbg!(&sig);
+    let result = builder.first_result(call);
+    builder.ins().ret(Some(result), span);
+    let function_id = builder.build()?;
+    module_builder.build()?;
+    let component_import = MidenAbiImport::new(FunctionType {
+        abi: Abi::Canonical,
+        params: import_lowered_sig.params.into_iter().map(|p| p.ty).collect(),
+        results: import_lowered_sig.results.into_iter().map(|r| r.ty).collect(),
+    });
+    // dbg!(&component_import);
+    // dbg!(&function_id);
+
+    Ok((component_import, function_id))
+}
+
+/// Rewrites the calls to the function `from` to the function `to` in the given function.
+/// Returns `true` if any call was rewritten.
+fn rewrite_calls(function: &mut Function, from: FunctionIdent, to: FunctionIdent) -> bool {
```

Comment:
Is this needed anymore? We shouldn't need to rewrite any calls if we're synthesizing the `canon lift` and `canon lower` declarations, since calls will already be referencing those. Or am I misunderstanding what this is handling?